extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java

```diff
@@ -148,15 +161,104 @@ void build(CombinedIndexBuildItem index,
+                        if (!modelClasses.contains(cn)) {
+                            if (requiresTransformation(res, i, modelClassNamesInternal)) {
+                                transformers.produce(new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer));
+                            }
+                        }
+                    }
                 }
             }
         }
     }
 
+    static final int CONSTANT_UTF8_TAG = 1;
+    static final int CONSTANT_INTEGER_TAG = 3;
+    static final int CONSTANT_FLOAT_TAG = 4;
+    static final int CONSTANT_LONG_TAG = 5;
+    static final int CONSTANT_DOUBLE_TAG = 6;
+    static final int CONSTANT_CLASS_TAG = 7;
+    static final int CONSTANT_STRING_TAG = 8;
+    static final int CONSTANT_FIELDREF_TAG = 9;
+    static final int CONSTANT_METHODREF_TAG = 10;
+    static final int CONSTANT_INTERFACE_METHODREF_TAG = 11;
+    static final int CONSTANT_NAME_AND_TYPE_TAG = 12;
+    static final int CONSTANT_METHOD_HANDLE_TAG = 15;
+    static final int CONSTANT_METHOD_TYPE_TAG = 16;
+    static final int CONSTANT_DYNAMIC_TAG = 17;
+    static final int CONSTANT_INVOKE_DYNAMIC_TAG = 18;
+    static final int CONSTANT_MODULE_TAG = 19;
+    static final int CONSTANT_PACKAGE_TAG = 20;
+
+    private boolean requiresTransformation(String res, ClassPathElement i, Set<String> modelInfo) {
+        ByteBuffer data = ByteBuffer.wrap(i.getResource(res).getData());
+        if (data.getInt() != 0xCAFEBABE) {
+            return false; //not a class file
+        }
+        data.getShort();//major
+        data.getShort();//minor
+        int constantPoolCount = data.getShort();
+        int currentCpInfoIndex = 1;
+        while (currentCpInfoIndex < constantPoolCount) {
+            currentCpInfoIndex++;
+            int cpInfoSize;
+            switch (data.get()) {
+                case CONSTANT_FIELDREF_TAG:
+                case CONSTANT_METHODREF_TAG:
+                case CONSTANT_INTERFACE_METHODREF_TAG:
+                case CONSTANT_INTEGER_TAG:
+                case CONSTANT_FLOAT_TAG:
+                case CONSTANT_NAME_AND_TYPE_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_DYNAMIC_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_INVOKE_DYNAMIC_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_LONG_TAG:
+                case CONSTANT_DOUBLE_TAG:
+                    cpInfoSize = 8;
+                    currentCpInfoIndex++;
+                    break;
+                case CONSTANT_UTF8_TAG:
+                    int strLength = 0xFFFF & data.getShort();
+                    cpInfoSize = 0;
+                    byte[] str = new byte[strLength];
+                    data.get(str);
+                    if (modelInfo.contains(new String(str, StandardCharsets.UTF_8))) {
```

Comment:
I guess this will also match string constants, since we're not looking at the `CONSTANT_CLASS_TAG` that points to this but probably fine because it's FQNs.